mlir/include/mlir/Dialect/Tensor/IR/TensorOps.td

```diff
@@ -121,6 +121,66 @@ def Tensor_CastOp : Tensor_Op<"cast", [
   let hasCanonicalizer = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// ConcatOp
+//===----------------------------------------------------------------------===//
+
+def Tensor_ConcatOp : Tensor_Op<"concat",
+    [Pure,
+     DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+     DeclareOpInterfaceMethods<ReifyRankedShapedTypeOpInterface>]> {
+  let summary = "tensor concatenation operation";
+  let description = [{
+    The "concat" operation constructs a tensor out of a variadic list of input
+    tensors, concatenated along a static dimension. All inputs and the result
+    type must share the same rank.
+
+    `dim` specifies the dimension along which to concatenate. The size of the
+    concatenated dimension in the result must be equal to the sum of the sizes
+    of the inputs along that dimension. All other dimensions in both the inputs
+    and result must be the same size.
+
+    Example:
+
+    ```mlir
+    %0 = tensor.concat dim(0) %0, %1, %2 :
```

Comment:
I was expecting the same as `tensor.pad` where there are ways to rewrite it as an equivalent set of ops in DPS (either as a single linalg op or as a group of linalg ops). Bufferization would either require the DPS rewrite in advance (although that linalg op is quite ugly), or we allocate based on the result shape and copy one-by-one into the allocated buffer.

I'm not married to leaving it as non-DPS. It's just more consistent with other input dialects and might make higher level transformations easier.

Reply:
After some more thought, if we ever wanted a `memref.concat`, I would much prefer having a destination on `tensor.concat` to get there. If others agree, I'm open to switching it to DPS and just providing nice helpers for constructing the destination (so that we don't duplicate that logic in every frontend lowering).

Reply:
I dont think `memref.concat` makes sense. The norm is that `memref` operations dont have copy semantics, so `memref.concat` might not be expressible as that (without introducing some pointer arithmetic stuff). I think at `memref` level `memref.subviews` + `memref.copy` are better representation of concat, with something else converting the `tensor.concat` into as much as in-place update as possible. Given that I think not having DPS for `concat` might be fine...

Reply:
I was assuming a `memref.concat` would take the destination and implicitly just be that sequence of subviews + copies. Maybe that is just a convenience op and not really worth adding, or `linalg.concat` would be preferred. The reason I'm thinking about wanting DPS now is if we wanted to tile then bufferize. Without the destination I would expect concat to turn into an alloc + sequence of copies. With a destination, we're concatenating into a slice.

Reply:
> I was assuming a `memref.concat` would take the destination and implicitly just be that sequence of subviews + copies. Maybe that is just a convenience op and not really worth adding, or `linalg.concat` would be preferred. The reason I'm thinking about wanting DPS now is if we wanted to tile then bufferize. Without the destination I would expect concat to turn into an alloc + sequence of copies. With a destination, we're concatenating into a slice.

ok, that is a fair point. We have two examples of both. One is `tensor.pad` that is not destination passing style and you have to do something special to turn it into destination passing style. We could do that, or we do what `tensor.pack` does which is destination passing style. On balance I would maybe go with DPS for what you suggest.

Reply:
No strong opinion from my side regarding DPS/non-DPS. For the sake of simplicity (and to some degree consistency with the tensor dialect; most ops in there are non-DPS), I'd slightly prefer a non-DPS variant that allocates when bufferizing it. We can lower the op to `tensor.empty` + multiple `tensor.insert_slice`, then use empty tensor elimination to infer the destination. Basically the same way we handle `tensor.pad` today.

I've been thinking of extending `bufferization.materialize_in_destination` with a region that can wrap arbitrary non-DPS op and provide the missing destination. That could be an alternative to turning more and more ops into DPS style.

Regarding `memref.concat`: I would just generate `memref.subview`+`memref.copy` directly to keep the memref dialect as simple as possible. (Unless you want to do tiling of memref IR.)